control-statistics/src/main/java/fi/nls/oskari/control/statistics/user/AddIndicatorDataHandler.java

```diff
@@ -0,0 +1,109 @@
+import fi.nls.oskari.control.statistics.plugins.StatisticalDatasourcePluginManager;
+import fi.nls.oskari.control.statistics.plugins.db.DatasourceLayer;
+import fi.nls.oskari.service.OskariComponentManager;
+import fi.nls.oskari.util.JSONHelper;
+import fi.nls.oskari.util.ResponseHelper;
+import org.json.JSONObject;
+import org.oskari.statistics.user.StatisticalIndicatorService;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@OskariActionRoute("AddIndicatorData")
+public class AddIndicatorDataHandler extends ActionHandler {
+
+    private StatisticalIndicatorService indicatorService;
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    @Override
+    public void init() {
+        super.init();
+        if (indicatorService == null) {
+            indicatorService = OskariComponentManager.getComponentOfType(StatisticalIndicatorService.class);
+        }
+    }
+
+    private static String PARAM_DATA = "data";
+
+    public void handleAction(ActionParameters params) throws ActionException {
+        params.requireLoggedInUser();
+
+        int datasourceId = params.getRequiredParamInt(StatisticsHelper.PARAM_DATASOURCE_ID);
+        StatisticalDatasourcePlugin datasource = StatisticalDatasourcePluginManager.getInstance().getPlugin(datasourceId);
+        if(datasource == null) {
+            throw new ActionParamsException("Invalid datasource:" + datasourceId);
+        }
+        if(!datasource.canModify(params.getUser())) {
+            throw new ActionDeniedException("User doesn't have permission to modify datasource:" + datasourceId);
+        }
+        String id = params.getRequiredParam(ActionConstants.PARAM_ID);
+        StatisticalIndicator indicator = datasource.getIndicator(params.getUser(), id);
+        if(indicator == null) {
+            // indicator removed or is not owned by this user
+            throw new ActionParamsException("Requested invalid indicator:" + id);
+        }
+
+        // This is just an addition to the existing model, NOT the whole model
+        JSONObject selectors = params.getHttpParamAsJSON(StatisticsHelper.PARAM_SELECTORS);
+        StatisticalIndicatorDataModel model = StatisticsHelper.getIndicatorDataModel(selectors);
+
+        // add the layer if it doesn't exist yet (only if linked to the datasource)
+        int regionsetId = params.getRequiredParamInt(StatisticsHelper.PARAM_REGIONSET);
+        DatasourceLayer layer = datasource.getSource().getLayers().stream()
+                .filter(x -> x.getMaplayerId() == regionsetId)
+                .findFirst()
+                .orElseThrow(() -> new ActionParamsException("Invalid regionset: " + regionsetId));
+        if(indicator.getLayer(layer.getMaplayerId()) == null) {
+            indicator.addLayer(layer);
+        }
+        try {
+            indicator.setDataModel(model);
+            Map<String, IndicatorValue> data = parseIndicatorData(params.getHttpParam(PARAM_DATA));
```

Comment:
Consider moving this line and the previous line outside of the `try-catch-all` block so that the ActionExceptions thrown by `parseIndicatorData(String data)` are thrown as-is.